### DISCUSSION TITLE
Multi-homed RPI and bad TLS certificate

### DISCUSSION THREAD

**Discussion by chrsoo** [k3s-io/k3s#6993] [General]
Hi,

I have a Raspberry PI CM4 module with Wifi, running DietPI v8.14.2.

After installing K3s using the following command:

```
curl -sfL https://get.k3s.io | sh -s - \
--write-kubeconfig-mode 644 \
--token REDACTED \
--bind-address 192.168.2.21 \
--advertise-address 192.168.2.21 \
--disable-cloud-controller --disable servicelb --disable local-storage
```
... the metrics server fails to start properly.

The following errors are being repeated in the logs:
```
Feb 19 16:39:11 node1 k3s[8871]: W0219 16:39:11.082206    8871 garbagecollector.go:752] failed to discover some groups: map[metrics.k8s.io/v1beta1:the server is currently unable to handle the request]
Feb 19 16:39:14 node1 k3s[8871]: I0219 16:39:14.909803    8871 log.go:198] http: TLS handshake error from 10.42.0.5:45198: remote error: tls: bad certificate
Feb 19 16:39:29 node1 k3s[8871]: I0219 16:39:29.906534    8871 log.go:198] http: TLS handshake error from 10.42.0.5:47260: remote error: tls: bad certificate
Feb 19 16:39:40 node1 k3s[8871]: E0219 16:39:40.951487    8871 resource_quota_controller.go:417] unable to retrieve the complete list of server APIs: metrics.k8s.io/v1beta1: the server is currently unable to handle the request
...
Feb 19 16:45:41 node1 k3s[8871]: W0219 16:45:41.992640    8871 garbagecollector.go:752] failed to discover some groups: map[metrics.k8s.io/v1beta1:the server is currently unable to handle the request]
Feb 19 16:45:44 node1 k3s[8871]: I0219 16:45:44.905281    8871 log.go:198] http: TLS handshake error from 10.42.0.5:57160: remote error: tls: bad certificate
Feb 19 16:45:53 node1 k3s[8871]: W0219 16:45:53.453530    8871 handler_proxy.go:105] no RequestInfo found in the context
Feb 19 16:45:53 node1 k3s[8871]: E0219 16:45:53.453688    8871 controller.go:113] loading OpenAPI spec for "v1beta1.metrics.k8s.io" failed with: Error, could not get list of group versions for APIService
Feb 19 16:45:53 node1 k3s[8871]: I0219 16:45:53.453762    8871 controller.go:126] OpenAPI AggregationController: action for item v1beta1.metrics.k8s.io: Rate Limited Requeue.
Feb 19 16:45:53 node1 k3s[8871]: W0219 16:45:53.455297    8871 handler_proxy.go:105] no RequestInfo found in the context
Feb 19 16:45:53 node1 k3s[8871]: E0219 16:45:53.455561    8871 controller.go:116] loading OpenAPI spec for "v1beta1.metrics.k8s.io" failed with: failed to retrieve openAPI spec, http error: ResponseCode: 503, Body
```

Please note the following

* Both the eth0 and wifi interfaces are up
* I intended to use the WIFI as a gateway for the cluster with worker nodes running on the local eth network 192.168.2.0/24.
* Without the `--advertise-address 192.168.2.21` argument k3s attempts to dial the dynamically allocated address on the WIFI interface. This fails as the API server is bound to 192.168.2.21.

**My assumption is that the metrics-server failure is caused by the TLS errors on the internal cluster network (10.42.0.0).**

* Is my understanding correct?
* Does anybody know how I can fix the certificate error?

**Comment by chrsoo** [comment]
Checking the pod logs, both the traefik and metric pod logged the same error:
```
Error from server: Get "https://192.168.2.21:10250/containerLogs/kube-system/metrics-server-5f9f776df5-p9jtz/metrics-server": x509: certificate is valid for 127.0.0.1, 192.168.1.114, not 192.168.2.21
```
... although the traefik pod starts up while the metrics-server never reaches the `ready` state.

Uninstalling and reinstalling with the `--tls-san 192.168.2.21` parameter does not make any difference.

**Reply by brandond** [reply]
I would recommend trying the --node-ip flag before going and setting a bunch of other addresses manually. I'm not sure we've tested a configuration where one of the cluster members is also acting as the NAT gateway for other members; you may run into odd issues with that due to how Kubernetes handles routing between pods and services.

**Reply by chrsoo** [reply]
That... actually works - thanks!

Using the following command to install k3s:
```
curl -sfL https://get.k3s.io | sh -s - \
--write-kubeconfig-mode 644 \
--token REDACTED \
--node-ip 192.168.2.21 \
--disable-cloud-controller --disable servicelb --disable local-storage
```
